felix/bpf/xdp/attach.go

```diff
@@ -200,9 +200,14 @@ func (ap *AttachPoint) DetachProgram() error {
 		return nil
 	}
 
-	ourProg, err := bpf.AlreadyAttachedProg(ap, path.Join(bpfdefs.ObjectDir, ap.FileName()), progID)
-	if err != nil || !ourProg {
-		return fmt.Errorf("XDP expected program ID does match with current one: %w", err)
+	prog, err := bpf.GetProgByID(progID)
+	if err != nil {
+		return fmt.Errorf("failed to get prog by id %d: %w", progID, err)
+	}
+
+	if prog.Name != "cali_xdp_preamb" {
```

Comment:
Added the prefix match

> If we're relying on specific names, need UTs that will catch someone doing a well-meaning rename to cali_xdp_prmble or something...

We overall need to address cleanup of previous versions. Some of the stuff I killed in this PR needs to be preserved and ran at felix restart. If we upgrade from N-1 to N in which this change lands, we need to clean up after N-1 otherwise the per-ep programs and maps will stay pinned :disappointed: I may add that as a separate PR.

Reply:
Added logic to remove per-endpoint maps (no more in use) and old programs, not the preamble ones.